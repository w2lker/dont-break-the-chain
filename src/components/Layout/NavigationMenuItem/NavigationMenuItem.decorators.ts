import withStyles from '@material-ui/core/styles/withStyles';

import NavigationMenuItem from './NavigationMenuItem';
import navigationMenuItemStyles from './NavigationMenuItem.styles';

// @ts-ignore
const styledComponent = withStyles(navigationMenuItemStyles)(NavigationMenuItem);

export default styledComponent;
