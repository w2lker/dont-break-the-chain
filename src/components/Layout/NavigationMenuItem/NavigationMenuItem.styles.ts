import { bon } from '../../../utils/styleUtils';

const navigationMenuItemStyles = {
  root: {
    color: '#979797',
    '&.Mui-selected': {
      color: '#1A80EF',
    },
  },
  label: {
    ...bon.setFont(12, 'inherit', 400, '0.4px'),
    lineHeight: '16px',
    paddingTop: 10,
  },
  icon: {
    ...bon.setFont(19, 'inherit', 'unset', 'unset'),
  },
};

export default navigationMenuItemStyles;
