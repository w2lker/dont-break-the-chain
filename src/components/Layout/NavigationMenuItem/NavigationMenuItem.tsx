import React from 'react';

import { BottomNavigationAction } from '@material-ui/core';
import { NavLink, NavLinkProps } from 'react-router-dom';
import MaterialDesignIcon from '../../basic/Material-Icon';

export interface INavMenuItemProps  {
  url: string;
  label: string;
  icon: string;
  // Decorator property
  classes?: any;
}

export class DecoratedNavLink extends React.Component<NavLinkProps> {
  // @ts-ignore
  displayName: 'DecoratedNavLink';
  render() {
    // @ts-ignore
    return (<NavLink {...this.props} activeClassName="Mui-selected" />);
  }
}

const NavigationMenuItem: React.FC<INavMenuItemProps> = (props) => {
  const { url, label, icon, classes } = props;
  const iconRender = (
    <MaterialDesignIcon
      className={classes.icon}
      name={icon}
    />
  );

  // @ts-ignore
  return (
    // @ts-ignore
    <BottomNavigationAction
      classes={{
        root: classes.root,
        label: classes.label,
      }}
      component={DecoratedNavLink}
      to={url}
      label={label}
      icon={iconRender}
      showLabel={true}
    />
  );
};

NavigationMenuItem.defaultProps = {
  url: '/',
  icon: '',
  classes: {},
};

export default NavigationMenuItem;
