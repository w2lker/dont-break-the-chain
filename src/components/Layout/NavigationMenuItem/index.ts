import NavigationMenuItem from './NavigationMenuItem.decorators';

export default NavigationMenuItem;
