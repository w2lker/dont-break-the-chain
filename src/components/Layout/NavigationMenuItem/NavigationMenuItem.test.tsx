import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { mount, render, shallow } from 'enzyme';

import NavigationMenuItem, { DecoratedNavLink, INavMenuItemProps } from './NavigationMenuItem';
import NavigationMenuItemDecorated from './NavigationMenuItem.decorators';
import navigationMenuItemStyles from './NavigationMenuItem.styles';

const classes = {
  root: 'nav-menu-item-root',
  label: 'nav-menu-item-label',
  icon: 'nav-menu-item-icon',
};

const classesKeys = Object.keys(classes);

describe('DecoratedNavLink component renders', () => {
  const definedActiveClass = 'Mui-selected';
  // @ts-ignore
  const componentWithRouter = mount(
    <MemoryRouter>
      <DecoratedNavLink to="some-url"/>
    </MemoryRouter>,
  );
  const navLink = componentWithRouter.find('NavLink');

  it('match snapshot', () => {
    expect(componentWithRouter.debug()).toMatchSnapshot();
  });

  it('navLink receive activeClass', () => {
    // @ts-ignore
    const activeClass = navLink.props().activeClassName;
    expect(activeClass).toBe(definedActiveClass);
  });
});

describe('NavigationMenuItem Component', () => {

  const propsSample: INavMenuItemProps = {
    classes,
    label: 'habits',
    url: '/habits',
    icon: 'flag-checkered',
  };

  it('match snapshot', () => {
    const wrappedComponent = mount(
      <MemoryRouter>
        <NavigationMenuItem {...propsSample}/>
      </MemoryRouter>,
    );
    expect(wrappedComponent.debug()).toMatchSnapshot();
  });

  describe('Renders with empty props', () => {
    const props: INavMenuItemProps = {
      classes,
      label: '',
      url: '/',
      icon: '',
    };
    const wrappedComponent = mount(
      <MemoryRouter>
        <NavigationMenuItem {...props}/>
      </MemoryRouter>,
    );

    it('pass NavLink wrapper to BottomNavigationAction component', () => {
      // Depend on BottomNavigationAction root class
      const navLink = wrappedComponent.find('DecoratedNavLink');
      expect(navLink.length).toBe(1);
      expect(navLink.hasClass('MuiBottomNavigationAction-root')).toBeTruthy();
    });

    it('pass MaterialDesignIcon component', () => {
      const icon = wrappedComponent.find('MaterialDesignIcon');
      expect(icon.length).toBe(1);
      expect(icon.hasClass(classes.icon)).toBeTruthy();
    });

    it('pass classes to BottomNavigationAction', () => {
      const bottomNavigation = wrappedComponent.find('WithStyles(ForwardRef(BottomNavigationAction))');
      expect(bottomNavigation.length).toBe(1);
      // @ts-ignore
      expect(bottomNavigation.props().classes).toMatchObject({
        root: classes.root,
        label: classes.label,
      });
    });
  });

  describe('Renders with sample props', () => {

    const wrappedComponent = mount(
      <MemoryRouter>
        <NavigationMenuItem {...propsSample}/>
      </MemoryRouter>,
    );

    it('passes NavLink wrapper to BottomNavigationAction component', () => {
      // Depend on BottomNavigationAction root class
      const navLink = wrappedComponent.find('DecoratedNavLink');
      expect(navLink.length).toBe(1);
    });

    it('passes icon to MaterialDesignIcon component', () => {
      const icon = wrappedComponent.find('MaterialDesignIcon');
      expect(icon.length).toBe(1);
      expect(icon.hasClass(classes.icon)).toBeTruthy();
      expect(icon.props().name).toBe(propsSample.icon);
    });

  });

});

describe('NavigationMenuItem Styles', () => {
  it('exists', () => {
    expect(navigationMenuItemStyles).toBeDefined();
  });
  classesKeys.forEach((keyValue) => {
    it(`has ${keyValue} class styles`, () => {
      // @ts-ignore
      expect(navigationMenuItemStyles[keyValue]).toBeDefined();
    });
  });
});

describe('NavigationMenuItem Decorated', () => {
  const props: INavMenuItemProps = {
    label: 'habits',
    url: '/habits',
    icon: 'flag-checkered',
  };

  const wrappedComponent = mount(
    <MemoryRouter>
      <NavigationMenuItemDecorated {...props}/>
    </MemoryRouter>,
  );
  const component = wrappedComponent.find('NavigationMenuItem');
  it('has root object', () => {
    expect(component.length).toBe(1);
  });
  // @ts-ignore
  const assignedClasses = component.props().classes;
  it('has classes map', () => {
    expect(assignedClasses).toBeDefined();
  });
  classesKeys.forEach((keyValue) => {
    it(`has ${keyValue} class attribute`, () => {
      expect(assignedClasses[keyValue]).toBeDefined();
    });
  });
});
