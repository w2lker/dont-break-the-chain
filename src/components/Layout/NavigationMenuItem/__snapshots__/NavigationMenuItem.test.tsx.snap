// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecoratedNavLink component renders match snapshot 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <DecoratedNavLink to=\\"some-url\\">
      <NavLink to=\\"some-url\\" activeClassName=\\"Mui-selected\\">
        <Link aria-current={{...}} className={[undefined]} style={[undefined]} to={{...}}>
          <LinkAnchor aria-current={{...}} className={[undefined]} style={[undefined]} href=\\"/some-url\\" navigate={[Function: navigate]}>
            <a aria-current={{...}} className={[undefined]} style={[undefined]} href=\\"/some-url\\" onClick={[Function: onClick]} />
          </LinkAnchor>
        </Link>
      </NavLink>
    </DecoratedNavLink>
  </Router>
</MemoryRouter>"
`;

exports[`NavigationMenuItem Component match snapshot 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <NavigationMenuItem classes={{...}} label=\\"habits\\" url=\\"/habits\\" icon=\\"flag-checkered\\">
      <WithStyles(ForwardRef(BottomNavigationAction)) classes={{...}} component={[function]} to=\\"/habits\\" label=\\"habits\\" icon={{...}} showLabel={true}>
        <ForwardRef(BottomNavigationAction) classes={{...}} component={[function]} to=\\"/habits\\" label=\\"habits\\" icon={{...}} showLabel={true}>
          <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiBottomNavigationAction-root nav-menu-item-root\\" focusRipple={true} onClick={[Function: handleChange]} component={[function]} to=\\"/habits\\">
            <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiBottomNavigationAction-root nav-menu-item-root\\" focusRipple={true} onClick={[Function: handleChange]} component={[function]} to=\\"/habits\\">
              <DecoratedNavLink className=\\"MuiButtonBase-root MuiBottomNavigationAction-root nav-menu-item-root\\" onBlur={[Function]} onClick={[Function: handleChange]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} role=\\"button\\" aria-disabled={false} to=\\"/habits\\">
                <NavLink className=\\"MuiButtonBase-root MuiBottomNavigationAction-root nav-menu-item-root\\" onBlur={[Function]} onClick={[Function: handleChange]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} role=\\"button\\" aria-disabled={false} to=\\"/habits\\" activeClassName=\\"Mui-selected\\">
                  <Link aria-current={{...}} className=\\"MuiButtonBase-root MuiBottomNavigationAction-root nav-menu-item-root\\" style={[undefined]} to={{...}} onBlur={[Function]} onClick={[Function: handleChange]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} role=\\"button\\" aria-disabled={false}>
                    <LinkAnchor aria-current={{...}} className=\\"MuiButtonBase-root MuiBottomNavigationAction-root nav-menu-item-root\\" style={[undefined]} onBlur={[Function]} onClick={[Function: handleChange]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} role=\\"button\\" aria-disabled={false} href=\\"/habits\\" navigate={[Function: navigate]}>
                      <a aria-current={{...}} className=\\"MuiButtonBase-root MuiBottomNavigationAction-root nav-menu-item-root\\" style={[undefined]} onBlur={[Function]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} role=\\"button\\" aria-disabled={false} href=\\"/habits\\" onClick={[Function: onClick]}>
                        <span className=\\"MuiBottomNavigationAction-wrapper\\">
                          <MaterialDesignIcon className=\\"nav-menu-item-icon\\" name=\\"flag-checkered\\">
                            <span className=\\"material-icon mdi mdi-flag-checkered nav-menu-item-icon\\" style={[undefined]} />
                          </MaterialDesignIcon>
                          <span className=\\"MuiBottomNavigationAction-label nav-menu-item-label\\">
                            habits
                          </span>
                        </span>
                        <NoSsr>
                          <WithStyles(undefined) center={false}>
                            <ForwardRef(TouchRipple) classes={{...}} center={false}>
                              <span className=\\"MuiTouchRipple-root\\">
                                <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                              </span>
                            </ForwardRef(TouchRipple)>
                          </WithStyles(undefined)>
                        </NoSsr>
                      </a>
                    </LinkAnchor>
                  </Link>
                </NavLink>
              </DecoratedNavLink>
            </ForwardRef(ButtonBase)>
          </WithStyles(ForwardRef(ButtonBase))>
        </ForwardRef(BottomNavigationAction)>
      </WithStyles(ForwardRef(BottomNavigationAction))>
    </NavigationMenuItem>
  </Router>
</MemoryRouter>"
`;
